### ISSUE TITLE
tf.contrib.quantize.create_eval_graph does not add min/max node to bias

### ISSUE THREAD

**Issue by ltrottier** [tensorflow/tensorflow#25768]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): `yes`
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): `Linux archlinux 4.19.2-arch1-1-ARCH #1 SMP PREEMPT Tue Nov 13 21:16:19 UTC 2018 x86_64 GNU/Linux`
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): `binary`
- TensorFlow version (use command below): `v1.12.0-0-ga6d8ffae09 1.12.0`
- Python version: `3.6.7`
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: `none`
- GPU model and memory: `none`

**Describe the current behavior**

I want to convert a model to TFLite with quantized uint8 as inference type. I add the min/max nodes to the graph by calling `tf.contrib.quantize.create_eval_graph`, then I load the trained values (both the weights and the min/max values). I try to convert the model with TFLiteConverter, but I get errors of this kind:

 `Array conv2d/BiasAdd does not have MinMax information, and is not a constant array. Cannot proceed with quantization.`

I look at my graph (the one modified by `tf.contrib.quantize.create_eval_graph`) and I don't see min/max nodes for biases.

**Describe the expected behavior**

The biases should have min/max nodes to be used by TFLiteConverter after calling `tf.contrib.quantize.create_eval_graph`.

**Code to reproduce the issue**

This example can reproduce this bug using a single Conv2d layer. Note that for the sake of keeping the example simple, I simulate loading the weights and the values of the min/max nodes (learned during training) by simply running `tf.global_variables_initializer()`. 

```python
import tensorflow as tf
import tensorflow.contrib.lite as tflite

sess = tf.keras.backend.get_session()

# create model
x = x_input = tf.keras.layers.Input([10,10,3])
x = tf.keras.layers.Conv2D(1, 3)(x)
model = tf.keras.Model(x_input, x)

# add fake quantization nodes 
tf.contrib.quantize.create_eval_graph(sess.graph)

# simulate loading weights and quantization min/max values
sess.run(tf.global_variables_initializer())

# try to quantize
toco_converter = tflite.TFLiteConverter.from_session(sess, model.inputs, model.outputs)
toco_converter.post_training_quantize = True
toco_converter.inference_type = tflite.constants.QUANTIZED_UINT8
toco_converter.inference_input_type = tflite.constants.QUANTIZED_UINT8
toco_converter.quantized_input_stats = {model.inputs[0].name.split(':')[0]: (0., 255.)}
model_tflite_binary = toco_converter.convert()
```
gives the following error:
```
2019-02-14 15:12:39.980124: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 8 operators, 13 arrays (0 quantized)
2019-02-14 15:12:39.980218: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 8 operators, 13 arrays (0 quantized)
2019-02-14 15:12:39.980322: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 2 operators, 5 arrays (1 quantized)
2019-02-14 15:12:39.980352: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 2 operators, 5 arrays (1 quantized)
2019-02-14 15:12:39.980371: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before quantization graph transformations: 2 operators, 5 arrays (1 quantized)
2019-02-14 15:12:39.980401: F tensorflow/contrib/lite/toco/graph_transformations/quantize.cc:129] Array conv2d/BiasAdd does not have MinMax information, and is not a constant array. Cannot proceed with quantization.
```

When I look at the content of tf.global_variables(),  I see nothing related to bias_quant: 
```
[<tf.Variable 'conv2d/kernel:0' shape=(3, 3, 3, 1) dtype=float32>,
 <tf.Variable 'conv2d/bias:0' shape=(1,) dtype=float32>,
 <tf.Variable 'conv2d/weights_quant/min:0' shape=() dtype=float32_ref>,
 <tf.Variable 'conv2d/weights_quant/max:0' shape=() dtype=float32_ref>,
 <tf.Variable 'conv2d/act_quant/min:0' shape=() dtype=float32_ref>,
 <tf.Variable 'conv2d/act_quant/max:0' shape=() dtype=float32_ref>]
```

Same thing when I look as the graph in tensorboard:
```python
with tf.summary.FileWriter('/tmp/conv') as fw:
    fw.add_graph(sess.graph)
```

![picture-of-tensorboard](https://user-images.githubusercontent.com/14935326/52816460-48a84080-306f-11e9-9cb2-e3fb2a044f58.png)

**Other info / logs**

Event file from FileWriter: [log.zip](https://github.com/tensorflow/tensorflow/files/2866713/log.zip)

**Comment by ltrottier** [comment]
As a side note, if I do `use_bias=False` in the Conv2d, no quantization node is added. The content of `tf.global_variables()` is simply `[<tf.Variable 'conv2d/kernel:0' shape=(3, 3, 3, 1) dtype=float32>]`.

**Comment by ved27** [comment]
I went through the source code of eval graph api,  fake layers don't get added at all the nodes.
fake layer gets added after (conv2d + bias) or (conv2d + relu) combo as act_quant/***

**Comment by ltrottier** [comment]
Yes, I noticed that. Also, you cannot use standalone "keras" because it does not work. I needed to stick with tf.keras. I could not use batch norm neither (from tf.keras or tf.layers). 

As a side note, I was able convert the mobilenet models from [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/models.md#image-classification-quantized-models), but the predictions from my .tflite models were different then theirs. I can't explain it.

Anyways, I was able to quantize a really simple Conv2D + ReLU network. I may try with 1.13 later.

**Comment by ved27** [comment]
Use batch_norm from tf.contrib.layers. This works as it considers fusedbatchnorm by default, which gets folded with the convolution before it when you call quantization apis like create_training_graph() or create_eval_graph().

When you use this batch norm, do not forget to call optimizer under the UPDATE_OPS scope

**Comment by ltrottier** [comment]
Is it possible to do this with keras API? I mean, can we add control dependencies when training with `model.fit_generator`?

**Comment by hahadashi** [comment]
> the

hi, i meet the problem that the batchnormalization can not get the min max value。how i set the UPDATE_OPS to the control dependencies when i train。can you give a demo？

**Comment by ved27** [comment]
Hi,

########
x = tf.placeholder(tf.float32, [None, 1], 'x')
y = tf.layers.batch_normalization(x, training=is_training)

update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
    y = tf.identity(y)
##########

will work.
Another way, you can create training step under " with tf.control_dependencies(update_ops):  "

---

**Comment by hahadashi** [comment]
my codes just try add the with tf.control_dependencies(update_ops): when update the gradients like 
*****
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
       with tf.control_dependencies(update_ops):
               self.train_op = optimizer.apply_gradients(avg_grads_var, global_step=global_step)
*****
is right？？

or
i need add “update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
y = tf.identity(y)”
in every back of bn layers？？

**Comment by hahadashi** [comment]
> Hi,
> 
> ########
> x = tf.placeholder(tf.float32, [None, 1], 'x')
> y = tf.layers.batch_normalization(x, training=is_training)
> 
> update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
> with tf.control_dependencies(update_ops):
> y = tf.identity(y)
> ##########
> 
> will work.
> Another way, you can create training step under " with tf.control_dependencies(update_ops): "

forgot to thanks for you first

**Comment by ved27** [comment]
Yeah, any one works

**Comment by hahadashi** [comment]
> Yeah, any one works

i also meet the problem 
"
Array kws_model/KWS_Model/tower_0/CNN_V1/first_bn/FusedBatchNorm_mul_0, which is an input to the Add operator producing the output array kws_model/KWS_Model/tower_0/CNN_V1/Relu, is lacking min/max data, which is necessary for quantization. Either target a non-quantized output format, or change the input graph to contain min/max information, or pass --default_ranges_min= and --default_ranges_max= if you do not care about the accuracy of results.\n
"
i have tried many methods to fix it. but didn't work.
===========================
only i remove the batch_normalization layers or i add converter.default_ranges_stats=(0,6). but i know these are not the true way.

**Comment by curioyang** [comment]
> > Yeah, any one works
> 
> # i also meet the problem
> "
> Array kws_model/KWS_Model/tower_0/CNN_V1/first_bn/FusedBatchNorm_mul_0, which is an input to the Add operator producing the output array kws_model/KWS_Model/tower_0/CNN_V1/Relu, is lacking min/max data, which is necessary for quantization. Either target a non-quantized output format, or change the input graph to contain min/max information, or pass --default_ranges_min= and --default_ranges_max= if you do not care about the accuracy of results.\n
> "
> i have tried many methods to fix it. but didn't work.
> only i remove the batch_normalization layers or i add converter.default_ranges_stats=(0,6). but i know these are not the true way.

Hey , Do you have any solution to deal the DEFAULT_RANGES_MIN/MAX now?
I tried to add MinMax information into my model use Post-training method by (/tensorflow/bazel-bin/tensorflow/lite/toco/toco) . 
But i don't know how to use the flage(	--arrays_extra_info_file=""      	string	Path to an optional file containing a serialized ArraysExtraInfo proto allowing to pass extra information about arrays not specified in the input model file, such as extra MinMax information.).
Please give me some suggestion. thx
